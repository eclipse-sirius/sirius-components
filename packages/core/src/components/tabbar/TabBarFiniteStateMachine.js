/*******************************************************************************
 * Copyright (c) 2018 Obeo and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2
 * which accompanies this distribution and is available at
 * https://www.eclipse.org/legal/epl-2.0.
 *******************************************************************************/

export const INITIAL__STATE = 'INITIAL__STATE';
export const PRISTINE__STATE = 'PRISTINE__STATE';
export const MOVED__STATE = 'MOVED__STATE';

export const INITIALIZE__ACTION = 'INITIALIZE__ACTION';
export const HANDLE_NEXT__ACTION = 'HANDLE_NEXT__ACTION';
export const HANDLE_PREVIOUS__ACTION = 'HANDLE_PREVIOUS__ACTION';
export const HANDLE_TAB_SELECTED__ACTION = 'HANDLE_TAB_SELECTED__ACTION';

export const FSM = {
  INITIAL__STATE: { INITIALIZE__ACTION: [PRISTINE__STATE] },
  PRISTINE__STATE: {
    HANDLE_NEXT__ACTION: [MOVED__STATE],
    HANDLE_PREVIOUS__ACTION: [MOVED__STATE],
    HANDLE_TAB_SELECTED__ACTION: [MOVED__STATE]
  },
  MOVED__STATE: {
    HANDLE_NEXT__ACTION: [MOVED__STATE],
    HANDLE_PREVIOUS__ACTION: [MOVED__STATE],
    HANDLE_TAB_SELECTED__ACTION: [MOVED__STATE]
  }
};
