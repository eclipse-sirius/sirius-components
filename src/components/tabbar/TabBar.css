/*******************************************************************************
 * Copyright (c) 2018 Obeo and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2
 * which accompanies this distribution and is available at
 * https://www.eclipse.org/legal/epl-2.0.
 *******************************************************************************/

.tabbar {
  display: grid;
  grid-template-columns: min-content 1fr min-content;

  border-top: 1px solid var(--light-smoke);
  border-bottom: 1px solid var(--light-smoke);

  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.tabbar-tabs {
  display: grid;
  grid-auto-rows: 1fr;
  grid-template-columns: repeat(auto-fit, minmax(10px, 1fr));
  grid-column-gap: var(--layoutDimension-m);
  padding: 0 var(--layoutDimension-m);
}

.tabbar-nav {
  padding: var(--layoutDimension-m) var(--layoutDimension-s);
  color: var(--blue);
  font-weight: var(--fontWeight-bold);
}
.tabbar-nav:first-child {
  border-right: 1px solid var(--light-smoke);
}
.tabbar-nav:last-child {
  border-left: 1px solid var(--light-smoke);
}
.tabbar-nav:hover {
  cursor: pointer;
  color: var(--white);
  background-color: var(--blue);
}
.tabbar-nav--disabled {
  color: var(--light-smoke);
}
.tabbar-nav--disabled:hover {
  cursor: default;
  color: var(--light-smoke);
  background-color: var(--white);
}
.tabbar-nav--contain-selection {
  color: var(--white);
  background-color: var(--light-blue);
}
