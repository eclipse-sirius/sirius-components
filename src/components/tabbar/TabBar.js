/*******************************************************************************
 * Copyright (c) 2018 Obeo and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2
 * which accompanies this distribution and is available at
 * https://www.eclipse.org/legal/epl-2.0.
 *******************************************************************************/

import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { classNames } from '../../common/classnames';

import { actionCreator, dispatcher } from './TabBarDispatcher';

import './TabBar.css';

const TABBAR__CLASS_NAMES = 'tabbar';
const TABBAR_NAV__CLASS_NAMES = 'tabbar-nav';
const TABBAR_NAV__DISABLED__CLASS_NAMES = 'tabbar-nav--disabled';
const TABBAR_TAB__CONTAIN_SELECTION__CLASS_NAMES = 'tabbar-nav--contain-selection';
const TABBAR_TABS__CLASS_NAMES = 'tabbar-tabs';
const TABBAR_TAB__CLASS_NAMES = 'tabbar-tab title-m';
const TABBAR_TAB_SELECTED__CLASS_NAMES = 'tabbar-tab--selected';

const propTypes = {
  tabs: PropTypes.array.isRequired,
  onTabClick: PropTypes.func.isRequired,
  selectedTabIndex: PropTypes.number.isRequired,
  numberOfTabsDisplayed: PropTypes.number.isRequired
};
const defaultProps = {
  tabs: [],
  onTabClick: () => {},
  selectedTabIndex: 0,
  numberOfTabsDisplayed: 3
};

/**
 * The TabBar component is used to display horizontally a list of elements with
 * two buttons to navigate into the list to the previous or next elements. It
 * also allows the selection of one specific element in the list.
 */
export class TabBar extends Component {
  constructor(props) {
    super(props);

    this.handleTabBarNavClick = this.handleTabBarNavClick.bind(this);
    this.handleTabClick = this.handleTabClick.bind(this);

    this.state = dispatcher(undefined, props, actionCreator.newInitializeAction());
  }

  dispatch(action, callback) {
    this.setState((prevState, props) => dispatcher(prevState, props, action), callback);
  }

  handleTabBarNavClick({ target }) {
    const side = target.getAttribute('data-side');
    if (side === 'previous' && this.state.isPreviousAvailable) {
      const action = actionCreator.newHandlePreviousAction();
      this.dispatch(action);
    } else if (side === 'next' && this.state.isNextAvailable) {
      const action = actionCreator.newHandleNextAction();
      this.dispatch(action);
    }
  }

  handleTabClick(event) {
    const { onTabClick } = this.props;
    const indexValue = event.target.getAttribute('data-index');
    const selectedTabIndex = parseInt(indexValue, 10);

    const action = actionCreator.newHandleTabSelectedAction(selectedTabIndex);
    this.dispatch(action, () => onTabClick(selectedTabIndex));
  }

  getTabBarNavClassNames(isAvailable, containsSelection) {
    let tabBarNavClassNames = TABBAR_NAV__CLASS_NAMES;
    if (!isAvailable) {
      tabBarNavClassNames = classNames(TABBAR_NAV__CLASS_NAMES, TABBAR_NAV__DISABLED__CLASS_NAMES);
    } else if (containsSelection) {
      tabBarNavClassNames = classNames(
        TABBAR_NAV__CLASS_NAMES,
        TABBAR_TAB__CONTAIN_SELECTION__CLASS_NAMES
      );
    }
    return tabBarNavClassNames;
  }

  renderTabs(tabs, startIndex, selectedTabIndex, numberOfTabsDisplayed) {
    const tabsToDisplay = [];
    const endIndex = Math.min(tabs.length, startIndex + numberOfTabsDisplayed);
    for (let index = startIndex; index < endIndex; index++) {
      let tab = tabs[index];

      let tabClassNames = TABBAR_TAB__CLASS_NAMES;
      if (index === selectedTabIndex) {
        tabClassNames = classNames(TABBAR_TAB__CLASS_NAMES, TABBAR_TAB_SELECTED__CLASS_NAMES);
      }
      tabsToDisplay.push(
        <div key={tab} data-index={index} className={tabClassNames} onClick={this.handleTabClick}>
          {tab}
        </div>
      );
    }
    return tabsToDisplay;
  }

  render() {
    const { className, tabs, numberOfTabsDisplayed } = this.props;
    const { index, isPreviousAvailable, isNextAvailable, selectedTabIndex } = this.state;

    const previousContainsSelection = selectedTabIndex < index;
    const nextContainsSelection = index + numberOfTabsDisplayed < selectedTabIndex + 1;

    const tabBarClassNames = classNames(TABBAR__CLASS_NAMES, className);
    const previousClassNames = this.getTabBarNavClassNames(
      isPreviousAvailable,
      previousContainsSelection
    );
    const nextClassNames = this.getTabBarNavClassNames(isNextAvailable, nextContainsSelection);
    const tabsToDisplay = this.renderTabs(tabs, index, selectedTabIndex, numberOfTabsDisplayed);

    return (
      <div className={tabBarClassNames}>
        <div
          className={previousClassNames}
          onClick={this.handleTabBarNavClick}
          data-side="previous">
          &lt;
        </div>
        <div className={TABBAR_TABS__CLASS_NAMES}>{tabsToDisplay}</div>
        <div className={nextClassNames} onClick={this.handleTabBarNavClick} data-side="next">
          &gt;
        </div>
      </div>
    );
  }
}
TabBar.propTypes = propTypes;
TabBar.defaultProps = defaultProps;
