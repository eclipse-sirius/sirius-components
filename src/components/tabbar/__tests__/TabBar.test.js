/*******************************************************************************
 * Copyright (c) 2018 Obeo and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2
 * which accompanies this distribution and is available at
 * https://www.eclipse.org/legal/epl-2.0.
 *******************************************************************************/

import React from 'react';
import ReactDOM from 'react-dom';
import ReactTestUtils from 'react-dom/test-utils';
import Renderer from 'react-test-renderer';

import { TabBar } from '../TabBar';

import { PRISTINE__STATE, MOVED__STATE } from '../TabBarFiniteStateMachine';

describe('TabBar', () => {
  it('renders a list of tabs', () => {
    const tabs = ['First', 'Second', 'Third'];
    const tabBarComponent = Renderer.create(<TabBar tabs={tabs} />);
    expect(tabBarComponent.toJSON()).toMatchSnapshot();

    const tabBar = tabBarComponent.getInstance();
    const { stateId } = tabBar.state;
    expect(stateId).toBe(PRISTINE__STATE);
  });

  it('renders a list of tabs with a selected tab', () => {
    const tabs = ['First', 'Second', 'Third'];
    const tabBarComponent = Renderer.create(<TabBar tabs={tabs} selectedTabIndex={1} />);
    expect(tabBarComponent.toJSON()).toMatchSnapshot();

    const tabBar = tabBarComponent.getInstance();
    const { stateId } = tabBar.state;
    expect(stateId).toBe(PRISTINE__STATE);
  });

  it('changes selected tab when clicked', () => {
    let selectedTabIndex = 0;

    const div = document.createElement('div');
    const tabs = ['First', 'Second', 'Third'];
    const tabBar = ReactDOM.render(
      <TabBar tabs={tabs} onTabClick={index => (selectedTabIndex = index)} />,
      div
    );

    expect(tabBar.state.selectedTabIndex).toBe(0);

    const tabBarDOMNode = div.childNodes[0];
    const tabsDOMNode = tabBarDOMNode.childNodes[1];
    const secondTabDOMNode = tabsDOMNode.childNodes[1];

    ReactTestUtils.Simulate.click(secondTabDOMNode);
    expect(selectedTabIndex).toEqual(1);
    expect(tabBar.state.selectedTabIndex).toBe(1);
  });

  it('changes the tab rendered when clicking on next', () => {
    const div = document.createElement('div');
    const tabs = ['First', 'Second', 'Third', 'Fourth'];
    const tabBar = ReactDOM.render(<TabBar tabs={tabs} />, div);

    const tabBarDOMNode = div.childNodes[0];
    const nextDOMNode = tabBarDOMNode.childNodes[2];

    expect(tabBar.state.stateId).toBe(PRISTINE__STATE);
    expect(tabBar.state.isPreviousAvailable).toBe(false);
    expect(tabBar.state.isNextAvailable).toBe(true);
    expect(tabBar.state.index).toBe(0);

    ReactTestUtils.Simulate.click(nextDOMNode);

    expect(tabBar.state.stateId).toBe(MOVED__STATE);
    expect(tabBar.state.isPreviousAvailable).toBe(true);
    expect(tabBar.state.isNextAvailable).toBe(false);
    expect(tabBar.state.index).toBe(1);
  });
});
