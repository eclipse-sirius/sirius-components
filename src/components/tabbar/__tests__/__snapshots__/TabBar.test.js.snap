// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabBar renders a list of tabs 1`] = `
<div
  className="tabbar"
>
  <div
    className="tabbar-nav tabbar-nav--disabled"
    data-side="previous"
    onClick={[Function]}
  >
    &lt;
  </div>
  <div
    className="tabbar-tabs"
  >
    <div
      className="tabbar-tab title-m tabbar-tab--selected"
      data-index={0}
      onClick={[Function]}
    >
      First
    </div>
    <div
      className="tabbar-tab title-m"
      data-index={1}
      onClick={[Function]}
    >
      Second
    </div>
    <div
      className="tabbar-tab title-m"
      data-index={2}
      onClick={[Function]}
    >
      Third
    </div>
  </div>
  <div
    className="tabbar-nav tabbar-nav--disabled"
    data-side="next"
    onClick={[Function]}
  >
    &gt;
  </div>
</div>
`;

exports[`TabBar renders a list of tabs with a selected tab 1`] = `
<div
  className="tabbar"
>
  <div
    className="tabbar-nav tabbar-nav--disabled"
    data-side="previous"
    onClick={[Function]}
  >
    &lt;
  </div>
  <div
    className="tabbar-tabs"
  >
    <div
      className="tabbar-tab title-m"
      data-index={0}
      onClick={[Function]}
    >
      First
    </div>
    <div
      className="tabbar-tab title-m tabbar-tab--selected"
      data-index={1}
      onClick={[Function]}
    >
      Second
    </div>
    <div
      className="tabbar-tab title-m"
      data-index={2}
      onClick={[Function]}
    >
      Third
    </div>
  </div>
  <div
    className="tabbar-nav tabbar-nav--disabled"
    data-side="next"
    onClick={[Function]}
  >
    &gt;
  </div>
</div>
`;
